### ISSUE TITLE
visualize_hierarchy mix up of labels / hover over

### ISSUE THREAD

**Issue by wrighematthey** [MaartenGr/BERTopic#923]
I've found that for some datasets I'm getting different hover over text from the labels when using the visualize_hierarchy function. Tracking up the tree, the hover over labels seem to be in the correct order but the axis labels don't match at the lowest level. I've tried both with and without updating the topic names and this does not appear to make a difference. I think that the two lists are somehow being ordered differently but I can't track down where this occurs.

**Comment by MaartenGr** [comment]
Thanks for sharing this. Can you create a reproducible example of this? Also, can you share your code, perhaps there is something going on there.

**Comment by wrighematthey** [comment]
I've just tried running this with the newsgroup example from the docs but with a different model and can see the same effect. I used the 'all-MiniLM-L12-v2' model from sentence transformers. I don't seem to be able to upload the html version with the hover over tips but an example can be seen in the attached image.
![image](https://user-images.githubusercontent.com/13377601/211842940-d5e73b5c-e5b5-498b-9430-721d58b04bf8.png)

**Comment by wrighematthey** [comment]
Example code that produces output with described error

`# %%
from bertopic import BERTopic
from sklearn.datasets import fetch_20newsgroups

# %%
from sentence_transformers import SentenceTransformer
from umap import UMAP

# %%
docs = fetch_20newsgroups(subset='all', remove=('headers','footers','quotes'))['data']

# %%
models = {'small':'all-MiniLM-L6-v2',
          'medium':'all-MiniLM-L12-v2',
          'large':'all-mpnet-base-v2'}

# %%
sentence_model = SentenceTransformer(models['medium'])
topic_model = BERTopic(embedding_model=sentence_model)

# %%
embeddings = sentence_model.encode(docs,show_progress_bar=True)

# %%
from sklearn.feature_extraction.text import CountVectorizer

# %%
topics,probs = topic_model.fit_transform(docs, embeddings=embeddings)

# %%
vectorizer_model = CountVectorizer(stop_words="english", ngram_range=(1, 5))
topic_model.update_topics(docs, vectorizer_model=vectorizer_model)

# %%
reduced_embeddings = UMAP(n_neighbors=10, n_components=2, min_dist=0.0, metric='cosine').fit_transform(embeddings)

# %%
doc_plot = topic_model.visualize_documents(docs,reduced_embeddings=reduced_embeddings)
doc_plot.show()

# %%
hierarchical_topics = topic_model.hierarchical_topics(docs)

# %%
htm_plot = topic_model.visualize_hierarchy(hierarchical_topics=hierarchical_topics,orientation='bottom', custom_labels=False)
htm_plot.show()

# %%
htm_plot.write_html('news_example.html')
`

**Comment by MaartenGr** [comment]
Strange, I'll have to test this a bit further but it might be a result of one of the dependencies being updated. Could you do a pip freeze and share the packages that you currently have in your environment?

**Comment by hbeelee** [comment]
@MaartenGr I'm having the exactly same issue with **.visualize_hierarchy** and **.get_topic_tree**. 
The keywords of the topic on the graph (as a result of **.visualize_hierarchy**) and the keywords that appears when I hover over the labels do not mach - as @wrighematthey explained above.
In addition, the hierarchy that appears on  **.visualize_hierarchy** and that on **.get_topic_tree** are also totally different.

**Comment by MaartenGr** [comment]
It took a while to figure out and I should do a bit more testing but I believe the visualization should work if you remove the following lines: 

https://github.com/MaartenGr/BERTopic/blob/06dcd47a019854185a39146ed693d16dc3a27651/bertopic/plotting/_hierarchy.py#L232-L234

Hopefully, this should serve as a quick fix for now. I might change/update the plotting backend at some point so this might just be a temporary fix but I am not sure yet if I want to switch over to something like Bokeh.

**Comment by hbeelee** [comment]
Thank you very much for your feedbacks @MaartenGr, but it seems the problem persists. Would you please look into it anytime soon? I don't mean to rush you at all, though.

**Comment by MaartenGr** [comment]
@hbeelee Just to be sure, did you manually remove the lines of the code in `_hierarchy.py` and used the updated visualization function as a result? For me, this process seems to be working.

**Comment by sharma-n** [comment]
Hi !
I had the same issue on my side as well, and going from `0.13.0` to `0.12.0` seems to fix it for me (even though manually doing the change in [_hierarchy.py](https://github.com/MaartenGr/BERTopic/blob/06dcd47a019854185a39146ed693d16dc3a27651/bertopic/plotting/_hierarchy.py#L232-L234) doesn't solve the issue).
For my case, it wasn't just the labels that were wrong, but the hierarchy visualized did not match the one returned by `topic_model.hierarchical_topics()`
Hope this helps!

**Comment by MaartenGr** [comment]
@sharma-n Could you create a reproducible example? With manually doing the change I mentioned, I cannot seem to reproduce the issue which makes it difficult to fix.

**Comment by sharma-n** [comment]
Hey @MaartenGr ,
I was using a private dataset so couldn't add that. However, I found another public one where the issue can be reproduced ([from Kaggle](https://www.kaggle.com/datasets/datatattle/covid-19-nlp-text-classification), and have attached the CSV file [Corona_NLP_test.csv](https://github.com/MaartenGr/BERTopic/files/10836005/Corona_NLP_test.csv). 

I use the following code to process the data:
```python
from sentence_transformers import SentenceTransformer
from umap import UMAP
from hdbscan import HDBSCAN
from bertopic import BERTopic

data= pd.read_csv('Corona_NLP_test.csv')

sentence_model = SentenceTransformer('sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2')
embeddings = sentence_model.encode(data.OriginalTweet, show_progress_bar=False)
embeddings = UMAP(n_neighbors=15, n_components=5, min_dist=0.0, metric='cosine', random_state=42).fit_transform(embeddings)
reduced_embeddings = UMAP(n_neighbors=10, n_components=2, min_dist=0.0, metric='cosine', random_state=42).fit_transform(embeddings)
hdbscan_model = HDBSCAN(min_cluster_size=341, min_samples=3)

topic_model = BERTopic(hdbscan_model=hdbscan_model ,n_gram_range=(1,3), min_topic_size=341)
topics, probs = topic_model.fit_transform(data.OriginalTweet, embeddings)
hierarchical_topics = topic_model.hierarchical_topics(data.OriginalTweet)
topic_model.visualize_documents(data.OriginalTweet, reduced_embeddings=reduced_embeddings)
```

Without any changes to version `0.13.0`, I get the following output (I used custom labels, if that is of importance):
![13_nochange](https://user-images.githubusercontent.com/31072466/221474391-e115ebc9-6cd8-4e6d-a1f0-9991b307ba67.jpg)
Based on the hovertext, you can see it doesn't correspond to what it should be (matching the top topic better than the current topic)
If I do the changes you mentioned, I get a different results (that is still not correct):
![13_afterchange](https://user-images.githubusercontent.com/31072466/221474450-bdffcc55-553d-4208-8ac9-cfa6b290757f.jpg)
Finally, with version `0.12.0` I do get the expected results:
![12](https://user-images.githubusercontent.com/31072466/221476066-4ddca19f-8235-4370-9dbe-22d3b8f46274.jpg)

**Comment by MaartenGr** [comment]
Thanks for sharing the example! I think there needs to be one additional chance that might solve the issue, namely additionally removing the following lines: 

https://github.com/MaartenGr/BERTopic/blob/d665d3f8d8c7c1736dc82b1df8839ced56a2adb6/bertopic/_bertopic.py#L869-L871

Based on what you mentioned, the main thing that changed between v0.12 and v0.13 was the introduction of a 1-D condensed distance matrix as that is what scipy expects as an input. However, and I have no clue why, this is what messes up the order in the visualization.

**Comment by zilch42** [comment]
Hi Maarten, removing those two lines from both _bertopic.py and _hierarchy.py worked for me with regards to #1063. I have tested both visualize_hierarchy and visualize_hierarchical_documents and both are looking sensible.

**Comment by elashrry** [comment]
Hello @MaartenGr , 
I have the same issue. Removing the two lines you mentioned from both `_bertopic.py` and `_hierarchy.py` will solve the problem of matching labels and hover annotations. However, the clustering itself might not be correct. I think it lies in these [two lines](https://github.com/plotly/plotly.py/blob/9050c4b19c87b637a9867cb6cdbf4d442b54fd42/packages/python/plotly/plotly/figure_factory/_dendrogram.py#L340-L341) from Plotly's API. 
```
d = distfun(X)
Z = linkagefun(d)
``` 
SciPy's API [mentions](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html) the first argument 

> The input y may be either a 1-D condensed distance matrix or a 2-D array of observation vectors.

So, in your code, you assume the distance function returns a symmetric 2-D array of pairwise distances while Plotly assumes it returns a 1-D array. 

So I think the solution could be either 
1. to make sure the `distance_function` returns a 1-D array before using Plotly's `create_dendrogram` function in the plotting module. https://github.com/MaartenGr/BERTopic/blob/d665d3f8d8c7c1736dc82b1df8839ced56a2adb6/bertopic/plotting/_hierarchy.py#L126

2. Or, add in the docstring that `distance_function` should return a 1-D array, and adapt the code for that, i.e. remove the lines you mentioned in `hierarchical_topics` `_bertopic.py` and also in `_get_annotations` in `_hierarchy.py`. 

I would be happy to help, just let me know :)

**Comment by MaartenGr** [comment]
@elashrry Thanks for digging into this. I think another option would be to check for 2D distances and convert them like I did [here](https://github.com/MaartenGr/BERTopic/blob/f220ddf14fd339f9731b6f35ad4493713f890d36/bertopic/_bertopic.py#L871) using `scipy.spatial.distance.squareform` might also fix this issue. However, that would require running a test of the distance function to see whether it returns 2D or condensed 1D shapes. 

The reason for suggesting the above is that the "hard" work is being done in BERTopic and that the user does not have to figure out what to pass as both will then be accepted.

**Comment by elashrry** [comment]
I am not sure this solves the real issue here. I think you are already doing that wherever you can, but the issue is that plotly doesn't do the same thing, so we end up with two different clustering. One from `hierarchical_topics`in `_bertopic.py` (and it is the same one in `_get_annotations` in `_hierarchy.py`) and another one from plotly's `create_dendrogram`.

If you want minimal change and it won't affect your API, you can add these lines before calling plotly's `create_dendrogram`. I just tested it and it works. 
```

    # convret distance_function so that it returns a 1-D array
    def distfun(X):
        X = distance_function(X)
        np.fill_diagonal(X, 0)
        return squareform(X)
```

and to be precise, I am assuming the correct clustering is the one we get from the follwing code: 
```
embeddings = topic_model.c_tf_idf_ # no filtering just for simplicity here

distance_function = lambda X: 1 - cosine_similarity(X)
linkage_function = lambda x: hierarchy.linkage(
    X, 'ward', optimal_ordering=True)

fig, ax = plt.subplots(figsize=(10, 15))
# compute distances
X = distance_function(embeddings)
# Make sure it is the 1-D condensed distance matrix with zeros on the diagonal
np.fill_diagonal(X, 0)
X = squareform(X)
# linkage
Z = hierarchy.linkage(X, 'ward', optimal_ordering=True)
dn = hierarchy.dendrogram(Z, orientation='right', ax=ax)
```

**Comment by elashrry** [comment]
Yes, that would be more robust. Let me know if I can help

**Comment by MaartenGr** [comment]
If you have the time, then a PR would be appreciated! Otherwise, I might find some time in the upcoming weeks.
